### ISSUE TITLE
Support metrics ingestion

### ISSUE THREAD

**Issue by GiviMAD** [opensearch-project/data-prepper#242]
Currently I can't ingest metrics from my nodejs application using the OpenTelemetry node sdk, data-prepper does not support it.
Will be awesome if data-prepper could process the metrics collected by the OpenTelemetry collector so they can be ingested by OpenSearch.
An issue for this exists on the OpenDistro repo https://github.com/opendistro-for-elasticsearch/data-prepper/issues/669

**Comment by KarstenSchnitter** [comment]
We started an internal PoC by implementing a plugin to support OpenTelemetry metrics ingestion. So far we have support for  the metric types Gauge and Sum. We followed closely the approach, that was taken with the trace support. So we also support resource and metric attributes. Currently we try to work out, how to integrate histograms. The main point is to provide an approach to visualise the histograms easily in the UI. Support for summaries and exemplars will come later.

We would like to share the approach with the community. Is there interest from your side? We would need to get some internal approvals before we can share the code. But we are confident, that this will not be a major issue.

**Comment by dlvenable** [comment]
@KarstenSchnitter , Thank you for reaching out and working on this PoC. I'm very excited to work with your team on this!

Ideally, you can contribute your work as a plugin to the existing project so that it sits alongside the trace and log ingestion plugins. It would help for you to outline your approach in more detail in this issue (or create a new one) and then create a PR.

Please be sure to read the [Contribution Guide](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md) first. Then also please review the [Developer Guide](https://github.com/opensearch-project/data-prepper/blob/main/docs/developer_guide.md) before creating a PR.

After this work is completed, you could also submit a blog post so that it is visible in the [OpenSearch News](https://opensearch.org/blog). This blog accepts external contributions via PR.

**Comment by KarstenSchnitter** [comment]
The OTLP metrics format can best be understood from the [protobuf definitions](https://github.com/open-telemetry/opentelemetry-proto/blob/8ba33cceb4a6704af68a4022d17868a7ac1d94f4/opentelemetry/proto/metrics/v1/metrics.proto) especially the [overview diagram]( https://github.com/open-telemetry/opentelemetry-proto/blob/8ba33cceb4a6704af68a4022d17868a7ac1d94f4/opentelemetry/proto/metrics/v1/metrics.proto#L92-L128). For Gauges and Sums we split the data point array and index one document per data point. To this point we add all metadata such as metric name, description, unit or attributes. Hence, this metadata will be duplicated in each document. This allows for searches on this metadata. We oriented this approach on the otel-trace-*-prepper plugins. Example documents would look like this:

```json
[{
    "startTime": "2021-12-03T08:07:15.669894Z",
    "time": "2021-12-08T09:27:33.823874Z",
    "kind": "gauge",
    "value": "24.871355673763745",
    "unit": "Bytes",
    "description": "userrequest-gauge",
    "name": "gauge",
    "serviceName": "userservice",
    "resource.attributes.service@version": "1.0.0",
    "resource.attributes.service@name": "userservice",
    "instrumentationLibrary.version": "1.0.0",
    "instrumentationLibrary.name": "spring-boot-metrics-demo"
},{
    "startTime": "2021-11-30T16:50:58Z",
    "time": "2021-12-08T09:27:57.346911100Z",
    "kind": "sum",
    "value": "3282071552",
    "unit": "By",
    "description": "Disk bytes transferred.",
    "name": "system.disk.io",
    "serviceName": null,
    "aggregationTemporality": "AGGREGATION_TEMPORALITY_CUMULATIVE",
    "monotonic": true,
    "metric.attributes.device": "vda1",
    "metric.attributes.direction": "write"
 }]
```
These examples are taken from a very early stage. But they outline the data format we are aiming for. In the meantime, we think to understand, which attributes are repeated at top-level by the otel-trace plugins, e.g. serviceName. These attributes are used in the Trace UI plugins, right? Since we target no specific UI plugin but use tsvd, we would not need this duplication for now. We also recognised, that the final attribute field should be separated by an "@". 

The Contribution Guide looks rather straight-forward. I have requested approval from SAP for that contribution. After this is granted, we will provide a PR.
